File: Striver/DP/Q40.py
# buy and sell stocks with transaction fee

# recursive solution
def maxProfit(prices, i, buyAllow, fee):
    if i == len(prices):
        return 0
    profit = 0
    if buyAllow:
        buy = maxProfit(prices, i+1, False, fee) - prices[i]
        notBuy = maxProfit(prices, i+1, True, fee)
        profit = max(buy, notBuy)
    else:
        sell = maxProfit(prices, i+1, True, fee) + prices[i] - fee
        notSell = maxProfit(prices, i+1, False, fee)
        profit = max(sell, notSell)

    return profit

# memoization solution
def maxProfitMem(prices, i, buyAllow, dp, fee):
    if i == len(prices):
        return 0
    if dp[i][buyAllow] != -1:
        return dp[i][buyAllow]
    profit = 0
    if buyAllow:
        buy = maxProfitMem(prices, i+1, False, dp, fee) - prices[i]
        notBuy = maxProfitMem(prices, i+1, True, dp, fee)
        profit = max(buy, notBuy)
    else:
        sell = maxProfitMem(prices, i+1, True, dp, fee) + prices[i] - fee
        notSell = maxProfitMem(prices, i+1, False, dp, fee)
        profit = max(sell, notSell)

    dp[i][buyAllow] = profit
    return profit

# tabulation solution
def maxProfitTab(prices, fee):
    n = len(prices)
    dp = [[0 for _ in range(2)] for _ in range(n+1)]
    dp[n][1] = dp[n][0] = 0

    for i in range(n-1, -1, -1):
        for j in range(2):
            if j == 1:
                buy = dp[i+1][0] - prices[i]
                notBuy = dp[i+1][1]
                dp[i][j] = max(buy, notBuy)
            else:
                sell = dp[i+1][1] + prices[i] - fee
                notSell = dp[i+1][0]
                dp[i][j] = max(sell, notSell)

    return dp[0][1]

# space optimization solution
def maxProfitSpace(prices, fee):
    n = len(prices)
    ahead = [0 for _ in range(2)]
    curr = [0 for _ in range(2)]
    ahead[1] = 0
    ahead[0] = 0

    for i in range(n-1, -1, -1):
        for j in range(2):
            if j == 1:
                buy = ahead[0] - prices[i]
                notBuy = ahead[1]
                curr[j] = max(buy, notBuy)
            else:
                sell = ahead[1] + prices[i] - fee
                notSell = ahead[0]
                curr[j] = max(sell, notSell)

        ahead[0] = curr[0]
        ahead[1] = curr[1]

    return ahead[1]

# space optimization with variables solution
def maxProfitSpaceVar(prices, fee):
    n = len(prices)
    aheadBuy = 0
    aheadNotBuy = 0

    for i in range(n-1, -1, -1):
        currBuy = max(aheadNotBuy - prices[i], aheadBuy)
        currNotBuy = max(aheadBuy + prices[i] - fee, aheadNotBuy)
        aheadBuy = currBuy
        aheadNotBuy = currNotBuy

    return aheadBuy


prices = [1, 3, 2, 8, 4, 9]
fee = 2
print(maxProfit(prices, 0, True, fee))
dp = [[-1 for _ in range(2)] for _ in range(len(prices)+1)]
print(maxProfitMem(prices, 0, True, dp, fee))
print(maxProfitTab(prices, fee))
print(maxProfitSpace(prices, fee))
print(maxProfitSpaceVar(prices, fee))

File: Striver/DP/Q40.java
package Striver.DP;

public class Q40 {
    // buy and sell stocks with transaction fee
    // you can buy and sell multiple times

    // recursive solution - Time: O(2^n) Space: O(n)
    public int maxProfit(int[] prices, int idx, int buy, int fee){
        if(idx == prices.length)
            return 0;
        
        if(buy == 1){
            int buying = maxProfit(prices, idx+1, 0, fee) - prices[idx];
            int notBuying = maxProfit(prices, idx+1, 1, fee);
            return Math.max(buying, notBuying);
        }
        else{
            int selling = maxProfit(prices, idx+1, 1, fee) + prices[idx] - fee;
            int notSelling = maxProfit(prices, idx+1, 0, fee);
            return Math.max(selling, notSelling);
        }
    }

    // memoization - Time: O(n^2) Space: O(n^2)
    public int maxProfitMem(int[] prices, int idx, int buy, int[][] dp, int fee){
        if(idx == prices.length)
            return 0;
        
        if(dp[idx][buy] != -1)
            return dp[idx][buy];
        
        if(buy == 1){
            int buying = maxProfitMem(prices, idx+1, 0, dp, fee) - prices[idx];
            int notBuying = maxProfitMem(prices, idx+1, 1, dp, fee);
            return dp[idx][buy] = Math.max(buying, notBuying);
        }
        else{
            int selling = maxProfitMem(prices, idx+1, 1, dp, fee) + prices[idx] - fee;
            int notSelling = maxProfitMem(prices, idx+1, 0, dp, fee);
            return dp[idx][buy] = Math.max(selling, notSelling);
        }
    }

    // tabulation - Time: O(n^2) Space: O(n^2)
    public int maxProfitTab(int[] prices, int fee){
        int n = prices.length;
        int[][] dp = new int[n+1][2];
        dp[n][0] = dp[n][1] = 0;

        for(int i = n-1; i >= 0; i--){
            for(int j = 0; j < 2; j++){
                if(j == 1){
                    int buying = dp[i+1][0] - prices[i];
                    int notBuying = dp[i+1][1];
                    dp[i][j] = Math.max(buying, notBuying);
                }
                else{
                    int selling = dp[i+1][1] + prices[i] - fee;
                    int notSelling = dp[i+1][0];
                    dp[i][j] = Math.max(selling, notSelling);
                }
            }
        }

        return dp[0][1];
    }

    // space optimization - Time: O(n^2) Space: O(n)
    public int maxProfitSpaceOpt(int[] prices, int fee){
        int n = prices.length;
        int[] ahead = new int[2];
        int[] curr = new int[2];
        ahead[0] = ahead[1] = 0;

        for(int i = n-1; i >= 0; i--){
            for(int j = 0; j < 2; j++){
                if(j == 1){
                    int buying = ahead[0] - prices[i];
                    int notBuying = ahead[1];
                    curr[j] = Math.max(buying, notBuying);
                }
                else{
                    int selling = ahead[1] + prices[i] - fee;
                    int notSelling = ahead[0];
                    curr[j] = Math.max(selling, notSelling);
                }
            }
            for(int j = 0; j < 2; j++)
                ahead[j] = curr[j];
        }

        return ahead[1];
    }

    //space optimization with variable - Time: O(n^2) Space: O(1)
    public int maxProfitSpaceOptVar(int[] prices, int fee){
        int n = prices.length;
        int aheadNotBuying = 0, aheadBuying = 0, currNotBuying, currBuying;

        for(int i = n-1; i >= 0; i--){
            currNotBuying = Math.max(aheadBuying + prices[i] - fee, aheadNotBuying);
            currBuying = Math.max(aheadNotBuying - prices[i], aheadBuying);
            aheadBuying = currBuying;
            aheadNotBuying = currNotBuying;
        }

        return aheadBuying;
    }

    public static void main(String[] args) {
        Q40 obj = new Q40();
        int[] prices = {1, 3, 2, 8, 4, 9};
        int fee = 2;
        System.out.println(obj.maxProfit(prices, 0, 1, fee));
        int[][] dp = new int[prices.length+1][2];
        for(int i = 0; i < dp.length; i++)
            for(int j = 0; j < dp[0].length; j++)
                dp[i][j] = -1;
        System.out.println(obj.maxProfitMem(prices, 0, 1, dp, fee));
        System.out.println(obj.maxProfitTab(prices, fee));
        System.out.println(obj.maxProfitSpaceOpt(prices, fee));
        System.out.println(obj.maxProfitSpaceOptVar(prices, fee));
    }
}
